### ISSUE TITLE
Add `#[inline]` to almost all of the public API for every crate

### ISSUE THREAD

**Issue by blaumeise20** [gosub-browser/gosub-engine#398]
It is a well known "problem" in Rust that automatic cross-crate inlining is not supported. Every function that is seen as good to inline by the programmer has to be annotated with `#[inline]` to basically allow the Rust compiler to do inlining across crate boundaries. If you look into the standard library almost every function is marked as inline, however I haven't really seen that here. Inlining can have a great impact on performance, so we should consider to add `#[inline]` to pretty much the entire public API of the crates.

This page has a good explanation on how to configure inlining: https://nnethercote.github.io/perf-book/inlining.html

**Comment by Sharktheone** [comment]
Inlining should be done automatically by the compiler even if it hasn't the "permission" because we have enabled fat lto and have set codegen units to one in #310. It's enabled in the root crate, so it should also affect other crates that live in `crates/`.

But I'm not sure of is how that affects potential embedders of the engine. So it might be still needed to mark public functions as `#[inline(always)`, but this is something I need to test.

**Comment by blaumeise20** [comment]
Yeah that's what I assumed. With "root crate" you mean the binaries? If not, then anywhere the crates are included the configuration of those root crates will count.
I'm not sure if it is a good idea to add `always` to the inline directive, this is really only for small internal helper functions. `#[inline]` itself is mostly enough because the Rust compiler is really smart in this case. Just look at the standard library, I don't think there is *any* `always` at all.

**Comment by Sharktheone** [comment]
The root crate is `gosub_engine` which also contains the binaries, yes. But probably it's still good to have this configuration in every crate in case someone only uses one of these crates.

The stdlib also uses `#[inline(always]` for example in the `Rc` file there are 81 inline calls and 11 of those are `#[inline(always]` Another example is `v8` they seem to only use `[inline(always)]`.

I generally also think it shouldn't really be a problem when (almost) every function is inlined. Fat lto should from what I have read also do that.

@jaytaph What do you think?

**Comment by jaytaph** [comment]
I think that since the new engine API has only a very small surface as it's almost entirely event driven. I don't think these event calls are on any hot path so that we would have a need to inline them. If we do, we only increase compile time and file bloat without any real benefits.

Internally we already use fat LTO so we already get inlining on the hot paths. If we find some other function during benchmarking, we can always inline them manually.
